instrumentctl/g9_driver.py

```diff
@@ -0,0 +1,274 @@
+    8: 'Dual channel violation'
+}
+
+usStatus = {
+    9: "Output Power Supply Error Flag",
+    10: "Safety I/O Terminal Error Flag",
+    13: "Function Block Error Flag"
+}
+
+class G9Driver:
+    def __init__(self, port=None, baudrate=9600, timeout=0.5, logger=None, debug_mode=False):
+        if port:
+            self.ser = serial.Serial(port, baudrate, parity=serial.PARITY_EVEN, stopbits=serial.STOPBITS_ONE, bytesize=serial.EIGHTBITS, timeout=timeout)
+        self.debug_mode = debug_mode
+        self.logger = logger
+        self.lastResponse = None
+        self.msgOptData = None
+        self.input_flags = None
+
+    def send_command(self):
+        if not self.is_connected():
+            raise ConnectionError("Seiral Port is Not Open.")
+        header = b'\x40\x00\x00\x0F\x4B\x03\x4D\x00\x01' 
+        data = b'\x00\x00\x00\x00'
+        reserve = b'\x00\x00'
+        self.msgOptData = data
+        checksum = self.calculate_checksum(header + data + reserve, 0 , len(header + data))
+        footer = b'\x2A\x0D' 
+        #TODO: add try and catch
+        self.ser.write(header + data + reserve+ checksum + footer)
+
+        self.response()
+
+    # used mainly for the check sum but can also be used to check for error flags
+    # needs an input of a byte string and the range of bytes that need to be sum
+    # will return the sum of the bytes in the a byte string in the form of b'\x12'
+    def calculate_checksum(self, byteString, startByte, endByte):
+        assert isinstance(byteString, bytes)
+        return sum(byteString[startByte:endByte + 1]).to_bytes(2, "big") 
+
+    # helper function to convert bytes to bits for checking flags
+    # not currently being used but many be helpful in the future for getting errors
+    def bytes_to_binary(self, byte_string):
+        return ''.join(format(byte, '08b') for byte in byte_string)
+    
+    
+    # this method is made to check the error flags, right not only checks the last 13 bits
+    # of a byte string
+    def check_flags13(self, byteString, norm = 1):
+        assert isinstance(byteString, bytes)
+        binary_string = self.bytes_to_binary(byteString)[-13:]
+        string_of_ones = '1' * 13
+        return binary_string == string_of_ones
+           
+
+
+        # assert isinstance(byteString, bytes)
+        # # this is for if we only need the last 13 bits (more or less hardcoding this 
+        # # just including the rest if it might be helpful in the future
+        # if int(self.bytes_to_binary(byteString)[-13:], 2) >= (13 * norm):
+        #     # all flags we care about are 1
+        #     return True
+        # else:
+        #     return False
+
+    def response(self):
+        if not self.is_connected():
+            raise ConnectionError("Serial Port is Not Open.")
+        
+        data = self.ser.read_until('b\r')
+        self.lastResponse = data
+
+
+        # Indexing such that we don't return an integer
+        if data[0:1] == '\x40':
```

Comment:
Slicing the data works and I also swapped the character symbols to the hexadecimal byte representations. I tested this via an online python interpreter.